devito/ir/clusters/algorithms.py

```diff
@@ -186,3 +194,100 @@ def guard(clusters):
+        if not subiters:
+            return clusters
+
+        # Collect the index access functions along `d`, e.g., `t + 1` where `t` is
+        # a SteppingDimension for `d = time`
+        mapper = DefaultOrderedDict(lambda: DefaultOrderedDict(set))
+        for c in clusters:
+            indexeds = [a.indexed for a in c.scope.accesses if a.function.is_Tensor]
+
+            for i in indexeds:
+                try:
+                    iaf = i.indices[d]
+                except KeyError:
+                    continue
+
+                # Sanity checks
+                sis = iaf.free_symbols & subiters
+                if len(sis) == 0:
+                    continue
+                elif len(sis) == 1:
+                    si = sis.pop()
+                else:
+                    raise InvalidOperator("Cannot use multiple SteppingDimensions "
+                                          "to index into a Function")
+                size = i.function.shape_allocated[d]
+                assert is_integer(size)
+
+                mapper[size][si].add(iaf)
+
+        # Construct the ModuloDimensions
+        mds = []
+        for size, v in mapper.items():
+            for si, iafs in list(v.items()):
+                # Offsets are sorted so that the semantic order (t0, t1, t2) follows
+                # SymPy's index ordering (t, t-1, t+1) afer modulo replacement so
+                # that associativity errors are consistent. This corresponds to
+                # sorting offsets {-1, 0, 1} as {0, -1, 1} assigning -inf to 0
+                siafs = sorted(iafs, key=lambda i: -np.inf if i - si == 0 else (i - si))
+
+                for iaf in siafs:
+                    name = '%s%d' % (si.name, len(mds))
+                    offset = uxreplace(iaf, {si: d.root})
+                    mds.append(ModuloDimension(name, si, offset, size, origin=iaf))
+
+        # Replacement rule for ModuloDimensions
+        def rule(size, e):
+            try:
+                return e.function.shape_allocated[d] == size
+            except (AttributeError, KeyError):
+                return False
+
+        # Reconstruct the Clusters
+        processed = []
+        for c in clusters:
+            # Apply substitutions to expressions
+            # Note: In an expression, there could be `u[t+1, ...]` and `v[t+1,
+            # ...]`, where `u` and `v` are TimeFunction with circular time
+            # buffers (save=None) *but* different modulo extent. The `t+1`
+            # indices above are therefore conceptually different, so they will
+            # be replaced with the proper ModuloDimension through two different
+            # calls to `xreplace_indices`
+            exprs = c.exprs
+            groups = as_mapper(mds, lambda d: d.modulo)
+            for size, v in groups.items():
+                mapper = {md.origin: md for md in v}
+
+                func = partial(xreplace_indices, mapper=mapper, key=partial(rule, size))
```

Comment:
How expensive is that? `partial` always makes me worried for some reason.

Reply:
I don't know exactly how long that piece of code takes (I don't see why it should be slow though, I think you're biased when it gets to `partial`?) but yours, more generally is still a good question -- "how about the new compilation time for these Operators"?

So I took TTI so=4 and ran it with master and in my branch. I get:
* master: 0.27s for the pad Operators, 4.5s for the main TTI Operator
* this branch: 0.29s for the pad Operators, 5s for the main TTI Operator

So it is a bit slower now, I could run a profiler to find out why

Reply:
EDIT: well there's a whole new Cluster-level pass, the `lower_dimensions`, which creates the ModuloDimension, rather than using that awful global cache attached to DerivedDimension, which btw was a hack (causing failures here) and therefore has now been dropped

Reply:
That's a quite small variance I'd say so it's fine.